flyteplugins/go/tasks/pluginmachinery/internal/webapi/core.go

```diff
@@ -66,24 +67,39 @@ func (c CorePlugin) GetProperties() core.PluginProperties {
 }
 
 func (c CorePlugin) Handle(ctx context.Context, tCtx core.TaskExecutionContext) (core.Transition, error) {
+	taskTemplate, err := tCtx.TaskReader().Read(ctx)
+	if err != nil {
+		return core.UnknownTransition, err
+	}
+
+	// Use the sync plugin to execute the task if the task template has the sync plugin flavor.
+	if taskTemplate.GetMetadata().GetRuntime().GetFlavor() == syncPlugin {
+		phaseInfo, err := c.p.(webapi.SyncPlugin).Do(ctx, tCtx)
```

Comment:
What would happen if the contract is broken? For example the tasks does not correctly set the flavour that is expected by the plugin?

Reply:
The task will be executed by AsyncPlugin.
It is implemented by flytekit this PR.
We will set the variable `use_sync_plugin` to True.
https://github.com/flyteorg/flytekit/pull/1822/files#diff-63787311eeb5747eb3170136f7c05abe69fe01da23d94222a39baf798273e687R20-R52
And then eventually set the flavor variable to here.
https://github.com/flyteorg/flytekit/pull/1822/files#diff-d123cf5b0acf27c386c1ceb74fd4b0de0775378fe1864c76908b5a334f58fffaR173-R440

Reply:
I think the agent plugin task will return an error.
As we know, flytepropeller will call the function `invokePlugin`, and if the plugin is agent service, it will send a grpc request to `agent server`, start by `pyflyte serve`.
I think we can write 2 things to help users understand how to use the sync and async plugins easily.

1. add `logger.errorf ` when the plugin doesn't execute successfully, tell users to check whether they use the correct ones or not.
2. write more annotation for the function and the `use_sync_plugin` variable in both flytekit and flyteplugins.

Do you think this is a good solution?
Or there's any other ideas?
I would like to improve it, thanks a lot for your time.

Reply:
Sure we should provide good ways for users to figured out the error.

Maybe a more concrete question, what happens if the cast `c.p.(webapi.SyncPlugin)` fails here?

Reply:
I didn't try it before, but I totally agree that we should add a cast to check if it is correct!
Maybe something like this. (please ignore the implementation error here)
```
plugin, err := c.p.(webapi.SyncPlugin)
if err != nil {
                       logger.Errorf("please check if the sync plugin interface is implemented or not")
			return core.UnknownTransition, err
		}
```

Reply:
Yeah, this is the direction to go, and there might be other things to consider as well. Generally, if there exists a contract we need to be more protective on the backend side to avoid a total failure and have a better isolation.